Firestore/tests/Unit/FirestoreSessionHandlerTest.php

```diff
@@ -0,0 +1,441 @@
+            $this->connection->reveal(),
+            $this->valueMapper->reveal(),
+            self::PROJECT,
+            self::DATABASE
+        );
+        $firestoreSessionHandler->open(self::SESSION_SAVE_PATH, self::SESSION_NAME);
+        $ret = $firestoreSessionHandler->read('sessionid');
+
+        $this->assertEquals('', $ret);
+    }
+
+    /**
+     * @expectedException PHPUnit_Framework_Error_Warning
+     */
+    public function testReadWithException()
+    {
+        $this->connection->beginTransaction(['database' => $this->dbName()])
+            ->shouldBeCalledTimes(1);
+        $this->connection->batchGetDocuments([
+            'database' => $this->dbName(),
+            'documents' => [$this->documentName()],
+            'transaction' => null,
+        ])
+            ->shouldBeCalledTimes(1)
+            ->willThrow((new ServiceException('')));
+        $firestoreSessionHandler = new FirestoreSessionHandler(
+            $this->connection->reveal(),
+            $this->valueMapper->reveal(),
+            self::PROJECT,
+            self::DATABASE
+        );
+        $firestoreSessionHandler->open(self::SESSION_SAVE_PATH, self::SESSION_NAME);
+        $ret = $firestoreSessionHandler->read('sessionid');
+
+        $this->assertEquals('', $ret);
+    }
+
+    public function testReadEntity()
+    {
+        $this->documents->current()
+            ->shouldBeCalledTimes(1)
+            ->willReturn([
+                'found' => [
+                    'createTime' => date('Y-m-d'),
+                    'updateTime' => date('Y-m-d'),
+                    'readTime' => date('Y-m-d'),
+                    'fields' => ['data' => 'sessiondata']
+                ]
+            ]);
+        $this->valueMapper->decodeValues(['data' => 'sessiondata'])
+            ->shouldBeCalledTimes(1)
+            ->willReturn(['data' => 'sessiondata']);
+        $this->connection->beginTransaction(['database' => $this->dbName()])
+            ->shouldBeCalledTimes(1);
+        $this->connection->batchGetDocuments([
+            'database' => $this->dbName(),
+            'documents' => [$this->documentName()],
+            'transaction' => null,
+        ])
+            ->shouldBeCalledTimes(1)
+            ->willReturn($this->documents->reveal());
+        $firestoreSessionHandler = new FirestoreSessionHandler(
+            $this->connection->reveal(),
+            $this->valueMapper->reveal(),
+            self::PROJECT,
+            self::DATABASE
+        );
+
+        $firestoreSessionHandler->open(self::SESSION_SAVE_PATH, self::SESSION_NAME);
+        $ret = $firestoreSessionHandler->read('sessionid');
+
+        $this->assertEquals('sessiondata', $ret);
+    }
+
+    public function testWrite()
+    {
+        $phpunit = $this;
+        $this->valueMapper->encodeValues(Argument::type('array'))
+            ->will(function ($args) use ($phpunit) {
+                $phpunit->assertEquals('sessiondata', $args[0]['data']);
```

Comment:
nit: `$this` is available in callables starting in PHP 5.4, so you don't need to provide the `$phpunit` variable. `$this->assertEquals()` will work fine.